### PR TITLE
Enable Asset updates on iOS

### PR DESCRIPTION
**Pull request by geof90**
This PR enables asset updates on iOS.

The SSZipArchive changes are a result of adding an unzipping library to the project source. @lostintangent and I discussed and concluded that it will be simpler to ship SSZipArchive together in this plugin instead of relying on another iOS dependency manager like CocoaPods on top of NPM so as to keep our plugin installation simple.

react-native 15.0 and above is required after this change.

react

### CODE REVIEW

**Comment by lostintangent** [comment]
This is probably a stupid question, but do we even need the aes directory? I thought we only needed the main SSZipArchive files and then minizip. That is what their GitHub repo seems to indicate and is how the 3rd party RN Zip module works.

**Comment by geof90** [comment]
@lostintangent no we probably dont need it, i just copied the latest version off github which included them, ill spend some time to go over what we need and what we dont

File: CodePushDownloadHandler.m [hunk 1]
```diff
@@ -1,10 +1,13 @@
 #import "CodePush.h"
 
-@implementation CodePushDownloadHandler
+@implementation CodePushDownloadHandler {
+    // Header chars used to determine if the file is a zip.
+    char header[4];
```
Comment (lostintangent) [inline CodePushDownloadHandler.m]: _header

File: CodePushPackage.m [hunk 2]
```diff
@@ -179,21 +207,122 @@ + (void)downloadPackage:(NSDictionary *)updatePackage
         return failCallback(error);
     }
     
-    NSString *downloadFilePath = [packageFolderPath stringByAppendingPathComponent:UpdateBundleFileName];
+    NSString * downloadFilePath = [self getDownloadFilePath];
+    NSString * bundleFilePath = [newPackageFolderPath stringByAppendingPathComponent:UpdateBundleFileName];
```
Comment (lostintangent) [inline CodePushPackage.m]: No space between the * and the variable name

File: CodePushPackage.m [hunk 3]
```diff
@@ -179,21 +207,122 @@ + (void)downloadPackage:(NSDictionary *)updatePackage
...
+                [SSZipArchive unzipFileAtPath:downloadFilePath
+                                toDestination:unzippedFolderPath];
+                [[NSFileManager defaultManager] removeItemAtPath:downloadFilePath
+                                                           error:&nonFailingError];
+                if (nonFailingError) {
+                    NSLog(@"Error deleting downloaded file: %@", nonFailingError);
+                    nonFailingError = nil;
+                }
+                
+                NSString * diffManifestFilePath = [unzippedFolderPath
```
Comment (lostintangent) [inline CodePushPackage.m]: NSString *diffManifestFilePath

File: SSZipArchive/Common.h [hunk 4]
```diff
@@ -0,0 +1,81 @@
+#ifndef SSZipCommon
```
Comment (dtivel) [inline SSZipArchive/Common.h]: This file looks like a modification of https://github.com/samsoffes/ssziparchive/blob/master/minizip/unzip.h.

Please follow OSS license requirements.

File: CodePushDownloadHandler.m [hunk 5]
```diff
@@ -42,6 +45,18 @@ -(void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLRespons
 }
 
 -(void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
+    if (self.receivedContentLength < 4) {
+        const char *bytes = [data bytes];
+        for (int i=0; i<[data length]; i++) {
```
Comment (dtivel) [inline CodePushDownloadHandler.m]: Fix spacing.

    for (int i = 0; i < [data length]; i++) {

File: CodePushPackage.m [hunk 6]
```diff
@@ -179,21 +207,122 @@ + (void)downloadPackage:(NSDictionary *)updatePackage
...
+                                                   stringByAppendingPathComponent:DiffManifestFileName];
+                
+                if ([[NSFileManager defaultManager] fileExistsAtPath:diffManifestFilePath]) {
+                    // Copy the current package to the new package.
+                    NSString* currentPackageFolderPath = [self getCurrentPackageFolderPath:&error];
```
Comment (lostintangent) [inline CodePushPackage.m]: NSString *currentPackageFolderPath

File: CodePushPackage.m [hunk 7]
```diff
@@ -179,21 +207,122 @@ + (void)downloadPackage:(NSDictionary *)updatePackage
...
+                    NSString *manifestContent = [NSString stringWithContentsOfFile:diffManifestFilePath
+                                                                          encoding:NSUTF8StringEncoding
+                                                                             error:&error];
+                    if (error) {
+                        failCallback(error);
+                        return;
+                    }
+                    
+                    NSData *data = [manifestContent dataUsingEncoding:NSUTF8StringEncoding];
+                    NSDictionary* manifestJSON = [NSJSONSerialization JSONObjectWithData:data
```
Comment (lostintangent) [inline CodePushPackage.m]: NSDictionary *manifestJSON

File: CodePushDownloadHandler.m [hunk 8]
```diff
@@ -42,6 +45,18 @@ -(void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLRespons
 }
 
 -(void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
+    if (self.receivedContentLength < 4) {
+        const char *bytes = [data bytes];
```
Comment (dtivel) [inline CodePushDownloadHandler.m]: Move line 49 to line 55 to limit its scope to where it's actually used.

File: CodePushPackage.m [hunk 9]
```diff
@@ -179,21 +207,122 @@ + (void)downloadPackage:(NSDictionary *)updatePackage
...
+                                               error:&error];
+                [[NSFileManager defaultManager] removeItemAtPath:unzippedFolderPath
+                                                           error:&nonFailingError];
+                if (nonFailingError) {
+                    NSLog(@"Error deleting downloaded file: %@", nonFailingError);
+                    nonFailingError = nil;
+                }
+                
+                NSString *relativeBundlePath = [self findMainBundleInFolder:newPackageFolderPath
+                                                                   error:&error];
```
Comment (dtivel) [inline CodePushPackage.m]: Formatting is erratic, sometimes aligned on \:, sometimes on aligned on method name, sometimes on nothing.

File: SSZipArchive/Common.h (hunk 4 above)
Comment (geof90) [inline SSZipArchive/Common.h]: I got the file from https://github.com/ZipArchive/ZipArchive/blob/master/SSZipArchive/Common.h, it is unmodified. Will add a link to the README where I copied the files from.

Comment (geof90) [inline SSZipArchive/Common.h]: Added the README here. https://github.com/Microsoft/react-native-code-push/blob/asset-update/SSZipArchive/README.md

**Comment by geof90** [comment]
@lostintangent I think I there's no harm in leaving the AES folder there, [this file](https://github.com/Microsoft/react-native-code-push/blob/asset-update/SSZipArchive/minizip/zip.c#L43) references it, I don't want to have to modify it. Plus, having it there may let us do things like support password protected/encrypted zips in the future. :)

**Comment by ajwhite** [comment]
Woulds this resolve https://github.com/facebook/react-native/issues/4549 (RE https://github.com/Microsoft/react-native-code-push/issues/84#issuecomment-161839376)

**Comment by lostintangent** [comment]
Yep this would resolve that issue. Sorry for the inconvenience!

File: Examples/CodePushDemoApp/android/app/react.gradle [hunk 10]
```diff
@@ -0,0 +1,75 @@
+def config = project.hasProperty("react") ? project.react : [];
+
+def bundleAssetName = config.bundleAssetName ?: "index.android.bundle"
+def entryFile = config.entryFile ?: "index.android.js"
+
+// because elvis operator
```
Comment (dtivel) [inline Examples/CodePushDemoApp/android/app/react.gradle]: Remove references to Elvis operator.

You're using the ternary operator below (line 8).

http://groovy-lang.org/operators.html#_elvis_operator

File: SSZipArchive/Common.h (hunk 4 above)
Comment (dtivel) [inline SSZipArchive/Common.h]: Great.  Thanks.

**Comment by dtivel** [comment]
Looks good.

File: Examples/CodePushDemoApp/android/app/react.gradle (hunk 10 above)
Comment (geof90) [inline Examples/CodePushDemoApp/android/app/react.gradle]: This was also generated code from running the `react-native upgrade` command

https://github.com/facebook/react-native/blob/4fc3e26cc8794e0d40d45abd23d065dd18de4e22/local-cli/generator-android/templates/src/app/react.gradle

**Comment by lostintangent** [comment]
LGTM!

**Pull request merged**